architecture/spmc.rst

```diff
@@ -0,0 +1,267 @@
+
+:``sp_init_all()``:		Initialise all SPs which have been added by the
+				``SP_PATHS`` compiler option and run them
+:``thread_ffa_msg_wait()``:	All SPs are loaded and started. A
+				``FFA_MSG_WAIT`` message is send to the Normal
+				World.
+
+
+Each SP is loaded into the system using ldelf and started. This is done using
+Ldelf and is based around the same process as loading the early TAs.
+All SPs are run after they are loaded and run until a ``FFA_MSG_WAIT`` is send
+by the SP.
+
+
+.. uml::
+
+        autoactivate on
+        secure_partition.c -> secure_partition.c:	sp_init_uuid
+        secure_partition.c -> secure_partition.c:	sp_open_session
+        secure_partition.c -> secure_partition.c:	find_sp
+        return
+        secure_partition.c -> secure_partition.c:	sp_create_session
+        return
+        secure_partition.c -> secure_partition.c:	ldelf_load_ldelf
+        return
+        secure_partition.c -> secure_partition.c:	ldlelf_init_with_ldelf
+        return
+        secure_partition.c -> secure_partition.c:	sp_init_set_registers
+        return
+        return
+        secure_partition.c -> secure_partition.c:	enter_sp
+        return
+        secure_partition.c -> secure_partition.c:	sp_msg_handler
+        return
+        return
+
+:``init_with_ldelf()``:		Load the SP
+:``sp_init_info()``:		Initialise the ``struct ffa_init_info``. The
+				``struct ffa_init_info`` is passed to the SP
+				during it first run.
+:``sp_init_set_registers()``:	Initialise the registers of the SP
+:``sp_msg_handler()``:		Handle the SPs FF-A message
+
+Once all SP are loaded and started we return to the SPMD and the Normal World is
+booted.
+
+
+SP message handling
+-------------------
+
+The SPMC is split into 2 main message handlers:
+
+:``thread_spmc_msg_recv()`` thread_spm.c:	Used to handle message coming
+						from the Normal World.
+:``sp_msg_handler()`` spmc_sp_handler.c:	Used to handle message where
+						the source or the destination
+						is a SP.
+
+When a ``FFA_MSG_SEND_DIRECT_REQ`` message is received by the spmc from the
+Normal World, a new kernel thread is started.
+The FF-A message is passed to the kernel thread and it will call the
+``sp_msg_handler()`` function.
+
+Whenever the SPMC (``sp_msg_handler()``) receives a message not intended for
+one of the SPs, it will exit the kernel thread and return to the Normal World
+passing the FF-A message.
+
+Currently only a ``FFA_MSG_SEND_DIRECT_REQ`` can be passed from the Normal
+World to a SP.
+
+.. uml::
+        :width: 800
+
+        skinparam backgroundcolor transparent
+        participant None_secure_world
+
+        box "S-EL3"
+        participant SPMD
+        end box
+
+        box "S-EL1"
+        participant thread_spmc
+        participant spmc_sp_handler
+        end box
+
+
+        box "S-EL0"
+        participant SP
+        end box
+
+        autoactivate on
+        None_secure_world -> SPMD: FFA_MSG_SEND_DIRECT_REQ \n <font color=red>SMC
```

Comment:
"Non-secure world" instead?

Reply:
I can't have space in the participant name but I used the 'as' to change the tittle of the participant to "Non-secure world"

Reply:
OK, thanks